hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/metadata/HazelcastRelMdPrunability.java

```diff
@@ -0,0 +1,177 @@
+            RelMetadataQuery mq
+    ) {
+
+        final HazelcastTable hazelcastTable = OptUtils.extractHazelcastTable(scan);
+        if (!(hazelcastTable.getTarget() instanceof PartitionedMapTable)) {
+            return emptyMap();
+        }
+
+        final PartitionedMapTable targetTable = hazelcastTable.getTarget();
+        Set<String> partitioningColumns;
+
+        if (targetTable.partitioningAttributes().isEmpty()) {
+            partitioningColumns = ImmutableSet.of(QueryPath.KEY);
+        } else {
+            final HashSet<String> partitioningFieldNames = new HashSet<>(targetTable.partitioningAttributes());
+            partitioningColumns = targetTable.keyFields()
+                    .filter(kf -> partitioningFieldNames.contains(kf.getPath().getPath()))
+                    .map(TableField::getName)
+                    .collect(Collectors.toSet());
+        }
+
+        final RexNode filter = hazelcastTable.getFilter();
+        if (!(filter instanceof RexCall)) {
+            return emptyMap();
+        }
+
+        final RexCall call = (RexCall) filter;
+        final var conditionExtractor = new PartitionStrategyConditionExtractor();
+        return conditionExtractor.extractCondition(targetTable, call, partitioningColumns);
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(
+            IndexScanMapPhysicalRel scan,
+            RelMetadataQuery mq
+    ) {
+        // TODO: Implement
+        return emptyMap();
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(Calc calc, RelMetadataQuery mq) {
+        HazelcastRelMetadataQuery query = HazelcastRelMetadataQuery.reuseOrCreate(mq);
+        return query.extractPrunability(calc.getInput());
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(Aggregate agg, RelMetadataQuery mq) {
+        // TODO: Implement
+        return emptyMap();
+    }
+
+    // It is done to support usage of this metadata query during opt phase.
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(RelSubset subset, RelMetadataQuery mq) {
+        HazelcastRelMetadataQuery query = HazelcastRelMetadataQuery.reuseOrCreate(mq);
+        RelNode rel = Util.first(subset.getBest(), subset.getOriginal());
+        return query.extractPrunability(rel);
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(Join rel, RelMetadataQuery mq) {
+        // For any bi-rel (Joins) we (temporarily) are not propagating prunability.
+        return Collections.emptyMap();
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(RelNode rel, RelMetadataQuery mq) {
+        // For any non-mentioned rels, we assume they are prunable and forwards prunability.
```

Comment:
this is wrong, see: `shouldNotForwardPrunability_whenOneBranchIsNotPrunable`. More reasonable default could be: if all branches are prunable, then return sum of prunability, otherwise the node is not prunable if at least one input is not prunable. OFC this can be different for some cases (eg. HashJoin).

Reply:
BTW, this will enable partition pruning for queries that may not yet support it because of lacking support in Jet (like UNION). We MAY NOT break any existing query: every query that worked before partition pruning must work after it is implemented, even with AttributePartitionStrategy.

Reply:
so IMO it would be better to enable pruning explicitly for specific RelNodes

Reply:
> More reasonable default could be: if all branches/inputs are prunable, then return sum of prunability, otherwise the node is not prunable if at least one input is not prunable. 

Done .

Reply:
how about: 
> IMO it would be better to enable pruning explicitly for specific RelNodes
?
Is there any risk of introducing problems?

Reply:
What is my intention: during the design we decided that we are preparing an MVP, and for MVP current approach is fine. Afterward, we may add support for specific operators like `Aggregate` or `Join` and the design of the metadata query perfectly fits the incremental enhancement approach. So, I expect no changes now and later implement a metadata query method in `HazelcastRelMdPrunability` for specific nodes.

Reply:
I wasted some time (fortunately not that much) to create test that proves my point:

```java
    @Test
    public void test_simpleKeyPrunedOrderBy() {
        assertRowsAnyOrder("SELECT this FROM test1 WHERE __key = ? AND this = 'v1'  order by __key", List.of(1), rows(1, "v1"));
    }
```

Fails because of `java.lang.AssertionError: This item should not be handled by this collector as the requested partitionId is not present`

Reply:
we cannot leave the default behavior also because it is not future-proof. If we add a new rel, we need to consider partition pruning for it, not blindly use default and hope that is works + add tests for that.

Reply:
Not sure if you did update your branch, it works perfectly
![image](https://github.com/hazelcast/hazelcast/assets/10107557/184f03bc-f861-4835-9875-4bae46a0773c)

Reply:
@ivanthescientist disabled pruning for `__key` queries, so that is why it works now. You can construct similar example for AttrPartStrategy

Reply:
Ah, see it. Sorry.

Reply:
https://github.com/hazelcast/hazelcast/pull/24813/commits/6ef5f315055062bd3ec6767c1794d9daca1adfda

Done.

Reply:
I can see that you updated Sort handling (`allToOne` key) in CreateDagVisitor - that should be enough